server/src/main/java/org/elasticsearch/snapshots/SnapshotsServiceUtils.java

```diff
@@ -0,0 +1,1094 @@
+                                shards.put(shardId, new SnapshotsInProgress.ShardSnapshotStatus(primaryNodeId, shardStatus.generation()));
+                                continue;
+                            } else if (shardRouting.primaryShard().initializing() || shardRouting.primaryShard().relocating()) {
+                                // Shard that we were waiting for hasn't started yet or still relocating - will continue to wait
+                                shards.put(shardId, shardStatus);
+                                continue;
+                            }
+                        }
+                    }
+                    // Shard that we were waiting for went into unassigned state or disappeared (index or shard is gone) - giving up
+                    snapshotChanged = true;
+                    logger.warn("failing snapshot of shard [{}] on node [{}] because shard is unassigned", shardId, shardStatus.nodeId());
+                    final SnapshotsInProgress.ShardSnapshotStatus failedState = new SnapshotsInProgress.ShardSnapshotStatus(
+                        shardStatus.nodeId(),
+                        SnapshotsInProgress.ShardState.FAILED,
+                        shardStatus.generation(),
+                        "shard is unassigned"
+                    );
+                    shards.put(shardId, failedState);
+                    knownFailures.put(shardSnapshotEntry.getKey(), failedState);
+                } else if (shardStatus.state().completed() == false && shardStatus.nodeId() != null) {
+                    if (nodes.nodeExists(shardStatus.nodeId())) {
+                        shards.put(shardId, shardStatus);
+                    } else {
+                        // TODO: Restart snapshot on another node?
+                        snapshotChanged = true;
+                        logger.warn("failing snapshot of shard [{}] on departed node [{}]", shardId, shardStatus.nodeId());
+                        final SnapshotsInProgress.ShardSnapshotStatus failedState = new SnapshotsInProgress.ShardSnapshotStatus(
+                            shardStatus.nodeId(),
+                            SnapshotsInProgress.ShardState.FAILED,
+                            shardStatus.generation(),
+                            "node left the cluster during snapshot"
+                        );
+                        shards.put(shardId, failedState);
+                        knownFailures.put(shardSnapshotEntry.getKey(), failedState);
+                    }
+                } else {
+                    shards.put(shardId, shardStatus);
+                }
+        }
+        if (snapshotChanged) {
+            return shards.build();
+        } else {
+            return null;
+        }
+    }
+
+    static <T> void failListenersIgnoringException(@Nullable List<ActionListener<T>> listeners, Exception failure, Logger logger) {
```

Comment:
This is the other change to add the `logger` from `SnapshotsService`, along with `completeListenersIgnoringException()` below.